File: ZAJECIA_1/ZAD9_10/Reprezentacja_grafu.java
package JAVA;
import java.io.*;
import java.util.*;

public class Reprezentacja_grafu
{
    private int liczbaWierzcholkow;
    private int liczbaKrawedzi;
    private int[][] macierzSasiedztwa;
    private List<List<Integer>> listaSasiedztwa;
    private List<int[]> listaKrawedzi;
    private int[][] macierzIncydencji;

    public Reprezentacja_grafu(String nazwaPliku) 
    {
        try(BufferedReader br = new BufferedReader(new FileReader(nazwaPliku))) 
        {
            String[] pierwszaLinia = br.readLine().split(" ");
            liczbaWierzcholkow = Integer.parseInt(pierwszaLinia[0]);
            liczbaKrawedzi = Integer.parseInt(pierwszaLinia[1]);

            //Inicjalizacja struktur danych
            macierzSasiedztwa = new int[liczbaWierzcholkow][liczbaWierzcholkow];
            listaSasiedztwa = new ArrayList<>();
            listaKrawedzi = new ArrayList<>();
            macierzIncydencji = new int[liczbaWierzcholkow][liczbaKrawedzi];

            for(int i = 0; i < liczbaWierzcholkow; i++) 
            {
                listaSasiedztwa.add(new ArrayList<>());
            }

            //Wczytywanie krawedzi z pliku
            for(int i = 0; i < liczbaKrawedzi; i++) 
            {
                String[] linia = br.readLine().split(" ");
                int v = Integer.parseInt(linia[0]) - 1;
                int w = Integer.parseInt(linia[1]) - 1;

                listaKrawedzi.add(new int[]{v, w});
                macierzSasiedztwa[v][w] = 1;
                listaSasiedztwa.get(v).add(w);
                macierzIncydencji[v][i] = -1;
                macierzIncydencji[w][i] = 1;
            }

        } 
        catch(IOException e) 
        {
            System.err.println("Blad: Nie mozna otworzyc pliku " + nazwaPliku);
            System.exit(1);
        }
    }

    public void wypiszMacierzSasiedztwa() 
    {
        System.out.println("Macierz sasiedztwa:");
        for(int[] wiersz : macierzSasiedztwa) 
        {
            for(int wartosc : wiersz) 
            {
                System.out.print(wartosc + " ");
            }
            System.out.println();
        }
    }

    public void wypiszListeSasiedztwa() 
    {
        System.out.println("\nLista sasiedztwa:");
        for(int i = 0; i < liczbaWierzcholkow; i++) 
        {
            System.out.print((i + 1) + ": ");
            for(int sasiad : listaSasiedztwa.get(i)) 
            {
                System.out.print((sasiad + 1) + " ");
            }
            System.out.println();
        }
    }

    public void wypiszListeKrawedzi() 
    {
        System.out.println("\nLista krawedzi:");
        for(int[] krawedz : listaKrawedzi) 
        {
            System.out.println("(" + (krawedz[0] + 1) + ", " + (krawedz[1] + 1) + ")");
        }
    }

    public void wypiszMacierzIncydencji() 
    {
        System.out.println("\nMacierz incydencji:");
        for(int[] wiersz : macierzIncydencji) 
        {
            for(int wartosc : wiersz) 
            {
                System.out.print(wartosc + "\t");
            }
            System.out.println();
        }
    }

    public void wypiszReprezentacje() 
    {
        wypiszMacierzSasiedztwa();
        wypiszListeSasiedztwa();
        wypiszListeKrawedzi();
        wypiszMacierzIncydencji();
    }

    public static void main(String[] args) 
    {
        String nazwaPliku = "dane.txt";
        Reprezentacja_grafu graf = new Reprezentacja_grafu(nazwaPliku);
        graf.wypiszReprezentacje();
    }
}

File: ZAJECIA_1/ZAD9_10/Reprezentacja_grafu.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

class Graf
{
private:
    int liczba_wierzcholkow;
    int liczba_krawedzi;
    vector<vector<int>> macierz_sasiedztwa;
    vector<vector<int>> lista_sasiedztwa;
    vector<pair<int, int>> lista_krawedzi;
    vector<vector<int>> macierz_incydencji;

public:
    Graf(string nazwa_pliku)
    {
        ifstream plik(nazwa_pliku);
        if (!plik)
        {
            cerr << "Blad: Nie mozna otworzyc pliku " << nazwa_pliku << endl;
            exit(1);
        }

        // Wczytanie liczby wierzcholkow i krawedzi
        plik >> liczba_wierzcholkow >> liczba_krawedzi;

        // Inicjalizacja macierzy sasiedztwa i listy sasiedztwa
        macierz_sasiedztwa.resize(liczba_wierzcholkow, vector<int>(liczba_wierzcholkow, 0));
        lista_sasiedztwa.resize(liczba_wierzcholkow);
        macierz_incydencji.resize(liczba_wierzcholkow, vector<int>(liczba_krawedzi, 0));

        int v;
        int w;
        for (int i = 0; i < liczba_krawedzi; i++)
        {
            plik >> v >> w;
            v--;
            w--;
            lista_krawedzi.push_back({v, w});
            macierz_sasiedztwa[v][w] = 1;
            lista_sasiedztwa[v].push_back(w);
            macierz_incydencji[v][i] = -1;
            macierz_incydencji[w][i] = 1;
        }
        plik.close();
    }

    // Wypisanie macierzy sasiedztwa
    void wypiszMacierzSasiedztwa()
    {
        cout << "\nMacierz sasiedztwa:\n";
        for (const auto &wiersz : macierz_sasiedztwa)
        {
            for (int wartosc : wiersz)
            {
                cout << wartosc << " ";
            }
            cout << endl;
        }
    }

    // Wypisanie listy sasiedztwa
    void wypiszListeSasiedztwa()
    {
        cout << "\nLista sasiedztwa:\n";
        for (int i = 0; i < liczba_wierzcholkow; i++)
        {
            cout << i + 1 << ": ";
            for (int sasiad : lista_sasiedztwa[i])
            {
                cout << sasiad + 1 << " ";
            }
            cout << endl;
        }
    }

    // Wypisanie listy krawedzi
    void wypiszListeKrawedzi()
    {
        cout << "\nLista krawedzi:\n";
        for (auto krawedz : lista_krawedzi)
        {
            cout << "(" << krawedz.first + 1 << ", " << krawedz.second + 1 << ")" << endl;
        }
    }

    // Wypisanie macierzy incydencji
    void wypiszMacierzIncydencji()
    {
        cout << "\nMacierz incydencji:\n";
        for (const auto &wiersz : macierz_incydencji)
        {
            for (int wartosc : wiersz)
            {
                cout << wartosc << "\t";
            }
            cout << endl;
        }
    }

    // Wywolanie wszystkich metod wypisujacych
    void wypiszReprezentacje()
    {
        wypiszMacierzSasiedztwa();
        wypiszListeSasiedztwa();
        wypiszListeKrawedzi();
        wypiszMacierzIncydencji();
    }
};

int main()
{
    string nazwa_pliku = "dane.txt";
    Graf graf(nazwa_pliku);
    graf.wypiszReprezentacje();
    return 0;
}